### PR TITLE
Add #[private] method decorator

### LINKED ISSUE #122: Support "#[self_call]" for calls
It should check whether the current account equals to the predecessor account.
```
    assert_eq!(env::current_account_id(), env::predecessor_account_id());
``` 

Most of the time when you implement a callback function, you have to lock it for calls from the contract itself. This method assert check is very common and probably should be supported by the SDK

### PR DESCRIPTION
**Pull request by evgenykuzyakov**
Based on #168 by @chain-police 

Introduced `#[private]` method decorator, that verifies `predecessor_account_id() == current_account_id()`.
NOTE: Usually, when a contract has to have a callback for a remote cross-contract call, this callback method should
only be called by the contract itself. It's to avoid someone else calling it and messing the state. Pretty common pattern
is to have an assert that validates that the direct caller (predecessor account ID) matches to the contract's account (current account ID).

Fixes #122

### CODE REVIEW

**Review by ailisp — APPROVED** [review]
looks useful! but
`#[private] pub fn ...` does looks a bit strange. Maybe `#[owner]` from eth world is a better name?

**Review by nearmax — APPROVED** [review]
Please add a compilation test.

**Review by frol — APPROVED** [review]
Beautiful!

**Review by olonho — APPROVED** [review]
Cool!

File: README.md [hunk 1]
```diff
@@ -122,6 +122,21 @@ pub fn my_method(&mut self) {
...
+is to have an assert that validates that the direct caller (predecessor account ID) matches to the contract's account (current account ID).
+Macro `#[private]` simplifies it, by making it a single line macro instead and improves readability.
+
+To declare a private method use `#[private]` decorator:
+```rust
+
+#[private]
+pub fn my_method(&mut self) {
+...
+}
```
Comment (willemneal) [inline README.md:136]:
```suggestion
}
/// Which is equivalent to

pub fn my_method(&mut self ) {
    if env::current_account_id() != env::predecessor_account_id() {
        near_sdk::env::panic("Method method is private".as_bytes());
    }
...
}
```

**Review by willemneal — APPROVED** [review]
LGTM

**Comment by willemneal** [comment]
> looks useful! but
> `#[private] pub fn ...` does looks a bit strange. Maybe `#[owner]` from eth world is a better name?

Technically the macro could add the `pub` to make it feel more private.

**Comment by frol** [comment]
> Technically the macro could add the pub to make it feel more private.

Great point! How about even dropping the need to have the #[private] marker at all?

**Comment by evgenykuzyakov** [comment]
> How about even dropping the need to have the #[private] marker at all?

This will make all methods to be exported under `near_bindgen` which will probably generate JSON schema when it's needed. Unless we can explain that people need to have a separate `impl` section for non exported methods, then it might be confusing.

**Pull request merged**